processing/src/main/java/io/druid/segment/StringDimensionHandler.java

```diff
@@ -0,0 +1,1609 @@
+    public ValueMatcher makeIndexingValueMatcher(
+        final Comparable matchValue,
+        final IncrementalIndexStorageAdapter.EntryHolder holder,
+        final int dimIndex
+    )
+    {
+      final String value = STRING_TRANSFORMER.apply(matchValue);
+      final int encodedVal = getEncodedValue(value, false);
+      final boolean matchOnNull = Strings.isNullOrEmpty(value);
+      if (encodedVal < 0 && !matchOnNull) {
+        return new BooleanValueMatcher(false);
+      }
+
+      return new ValueMatcher()
+      {
+        @Override
+        public boolean matches()
+        {
+          Object[] dims = holder.getKey().getDims();
+          if (dimIndex >= dims.length) {
+            return matchOnNull;
+          }
+
+          int[] dimsInt = (int[]) dims[dimIndex];
+          if (dimsInt == null || dimsInt.length == 0) {
+            return matchOnNull;
+          }
+
+          for (int i = 0; i < dimsInt.length; i++) {
+            if (dimsInt[i] == encodedVal) {
+              return true;
+            }
+          }
+          return false;
+        }
+      };
+    }
+
+    @Override
+    public ValueMatcher makeIndexingValueMatcher(
+        final DruidPredicateFactory predicateFactory,
+        final IncrementalIndexStorageAdapter.EntryHolder holder,
+        final int dimIndex
+    )
+    {
+      final Predicate<String> predicate = predicateFactory.makeStringPredicate();
+      final boolean matchOnNull = predicate.apply(null);
+      return new ValueMatcher()
+      {
+        @Override
+        public boolean matches()
+        {
+          Object[] dims = holder.getKey().getDims();
+          if (dimIndex >= dims.length) {
+            return matchOnNull;
+          }
+
+          int[] dimsInt = (int[]) dims[dimIndex];
+          if (dimsInt == null || dimsInt.length == 0) {
+            return matchOnNull;
+          }
+
+          for (int i = 0; i < dimsInt.length; i++) {
+            String finalDimVal = getActualValue(dimsInt[i], false);
+            if (predicate.apply(finalDimVal)) {
+              return true;
+            }
+          }
+          return false;
+        }
+      };
+    }
+
+    private void updateSortedLookup()
+    {
+      if (sortedLookup == null) {
+        sortedLookup = dimLookup.sort();
+      }
+    }
+
+    private String getActualValue(int intermediateValue, boolean idSorted)
+    {
+      if (idSorted) {
+        updateSortedLookup();
+        return sortedLookup.getValueFromSortedId(intermediateValue);
+      } else {
+        return dimLookup.getValue(intermediateValue);
+
+      }
+    }
+
+    private int getEncodedValue(String fullValue, boolean idSorted)
+    {
+      int unsortedId = dimLookup.getId(fullValue);
+
+      if (idSorted) {
+        updateSortedLookup();
+        return sortedLookup.getSortedIdFromUnsortedId(unsortedId);
+      } else {
+        return unsortedId;
+      }
+    }
+  }
+
+
+  public class StringDimensionMerger implements DimensionMergerV9
```

Comment:
I had kept them totally separate initially to allow for easier deletion of StringDimensionMergerLegacy when the legacy IndexMerger is deprecated and removed.

Restructured this so that:
- DimensionMerger is base class with common methods for both V9 and legacy
- DimensionMergerV9 has V9 specific methods
- DimensionMergerLegacy has legacy specific methods
- StringDImensionMergerV9 is the base class for String implementation, StringDimensionMergerLegacy extends StringDimensionMergerV9 and overrides as needed, and also implements DimensionMergerLegacy. StringDimensioMergerLegacy can still be easily removed from the codebase when needed.